File: tests/Tests.FSharp/Sha256/Sha256Tests.fs
module Zeta.Tests.FSharp.Sha256.Sha256Tests

open System.Text
open Xunit
open Zeta.Core.FSharp.Sha256

[<Fact>]
let ``sha256 of empty string`` () =
    let input = Encoding.UTF8.GetBytes ""
    let hex = sha256Hex input
    Assert.Equal("e3b0c44298fc1c149afbf4c8996fb924" +
                 "27ae41e4649b934ca495991b7852b855", hex)

[<Fact>]
let ``sha256 of abc`` () =
    let input = Encoding.UTF8.GetBytes "abc"
    let hex = sha256Hex input
    Assert.Equal("ba7816bf8f01cfea414140de5dae2223" +
                 "b00361a396177a9cb410ff61f20015ad", hex)

[<Fact>]
let ``sha256 of nist two-block`` () =
    let input = Encoding.UTF8.GetBytes "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
    let hex = sha256Hex input
    Assert.Equal("248d6a61d20638b8e5c026930c3e6039" +
                 "a33ce45964ff2167f6ecedd419db06c1", hex)

[<Fact>]
let ``sha256 returns 32 bytes`` () =
    let input = Encoding.UTF8.GetBytes "test"
    let hash = sha256 input
    Assert.Equal(32, hash.Length)

File: tests/Tests.CSharp/Sha256/Sha256Tests.cs
using System.Text;
using Xunit;
using Sha256Lib = Zeta.Core.CSharp.Sha256.Sha256;

namespace Zeta.Tests.CSharp.Sha256;

/// <summary>
/// Standard-anchored SHA-256 tests (FIPS 180-4 / NIST test vectors).
/// These pass through System.Security.Cryptography.SHA256 via our BCL-clean wrapper.
/// </summary>
public class Sha256Tests
{
    // FIPS 180-4 / NIST SHA-256 vector: empty string.
    [Fact]
    public void EmptyStringMatchesNist()
    {
        var hex = Sha256Lib.HashHex(Encoding.UTF8.GetBytes(""));
        Assert.Equal("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", hex);
    }

    // FIPS 180-4 / NIST SHA-256 vector: "abc".
    [Fact]
    public void AbcMatchesNist()
    {
        var hex = Sha256Lib.HashHex(Encoding.UTF8.GetBytes("abc"));
        Assert.Equal("ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad", hex);
    }

    // FIPS 180-4 / NIST SHA-256 vector: two-block message (448-bit boundary).
    [Fact]
    public void NistTwoBlockMatchesNist()
    {
        var hex = Sha256Lib.HashHex(Encoding.UTF8.GetBytes("abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"));
        Assert.Equal("248d6a61d20638b8e5c026930c3e6039a33ce45964ff2167f6ecedd419db06c1", hex);
    }

    // Hash must produce exactly 32 bytes.
    [Fact]
    public void HashReturns32Bytes()
    {
        var bytes = Sha256Lib.Hash(Encoding.UTF8.GetBytes("abc"));
        Assert.Equal(32, bytes.Length);
    }
}
